hydra-node/src/Hydra/Persistence.hs

```diff
@@ -49,12 +50,190 @@ createPersistence fp = do
+  -- more complicated solutions would be possible, in particular, rolling hash / merkle chain might be more resilient to corruption
+  -- but given that persistence is already atomic and only needs to be consistent within a single node, it should suffice
+  nextId <- newTVarIO (0 :: Word64)
+  let eventSource =
+        EventSource
+          { getEvents' = do
+              tid <- myThreadId
+              atomically $ do
+                authTid <- readTVar authorizedThread
+                when (isJust authTid && authTid /= Just tid) $
+                  throwSTM (IncorrectAccessException $ "Trying to load persisted data in " <> fp <> " from different thread")
+
+              liftIO (doesFileExist fp) >>= \case
+                False -> pure []
+                True -> do
+                  bs <- readFileBS fp
+                  -- NOTE: We require the whole file to be loadable. It might
+                  -- happen that the data written by 'append' is only there
+                  -- partially and then this will fail (which we accept now).
+                  result <- case forM (C8.lines bs) Aeson.eitherDecodeStrict' of
+                    Left e -> throwIO $ PersistenceException e
+                    Right decoded -> pure decoded
+                  -- set initial nextId (zero-indexed) based on how many state change events we have
+                  atomically $ do
+                    writeTVar lastStateChangeId $ fromIntegral $ length result
+                    writeTVar nextId . fromIntegral $ length result
+
+                  pure result
+          }
+      eventSink =
+        EventSink
+          { putEvent' = \a -> do
+              threadId <- myThreadId
+              isEventNew <- atomically $ do
+                let stateChangeID = getID a
+
+                let outgoingStateChangeId = stateChangeID
+                -- outgoingStateChangeId <- readTVar $ stateChangeID -- this is the ID of the state change we just got from the node
+                -- it's not actually written to disk yet until this function is over
+
+                id <- readTVar nextId
+                writeTVar authorizedThread $ Just threadId
+                modifyTVar' nextId succ
+                pure $ outgoingStateChangeId `compare` id
+              let bytes = toStrict $ Aeson.encode a <> "\n"
+              case isEventNew of
+                -- event already persisted
+                LT -> pure ()
+                -- event is as new as expected
+                EQ -> liftIO $ withBinaryFile fp AppendMode (`BS.hPut` bytes)
+                -- event is newer than expected,
+                GT -> do
+                  liftIO $ putStrLn "ELAINE: this shouldn't happen with my current understanding of stuff"
+                  liftIO $ withBinaryFile fp AppendMode (`BS.hPut` bytes) -- FIXME(Elaine): maybe error ? shouldnt really happen
+          }
+      eventSinks = eventSink :| []
+  pure (eventSource, eventSinks)
+
```

Comment:
Should: just use `eventPairFromPersistenceIncremental`?

I am not sure why you copied/vendored/re-implemented these two functions above? Defining an `EventSource` and `EventSink` in terms of `PersistenceIncremental` is the way to go for this PR!